pkg/common/telemetry/version.go

```diff
@@ -0,0 +1,11 @@
+package telemetry
+
+import (
+	"github.com/spiffe/spire/pkg/common/version"
+)
+
+func EmitVersion(m Metrics) {
+	m.SetGaugeWithLabels([]string{"version"}, 1, []Label{
```

Comment:
It feels a little odd to have a metric with constant value, where the name (and usefulness) of the metric is solely fulfilled by a label. What are your thoughts on making this a counter with a name like "agent_boot" and "server_boot" or something along those lines?

Reply:
I suppose just `boot` or `started` could suffice too since we already prefix with agent or server.

Reply:
Yeah I get the confusion, the value here is a little moot. I'm not sure a counter like you're suggesting is any clearer since the counter value would be reset anytime the server or agent was restarted, effectively making it a constant value of 1 for each active instance. Perhaps a middle-ground solution of calling this metric "{agent,server}_boot" and keeping it as a gauge with value 1 would be clearer? It seems a little more intuitive to me at least that a gauge (as opposed to a counter) might have a constant value for an extended period of time.

Reply:
The value-prop of a value of 1 is that you can use an aggregation function (such as `sum`) to get the count of instances, rather than "counting" the series.

Off the top of my head I'm not sure there exists a query language (such as promql, graphite, m3ql, etc.) that allows you to "count" series grouped by tags. For example, if I have the ability to query global metrics, and I want a count of running version by-deployment-area I think I need to use a sum / groupBy function, rather than a "count of series".

Reply:
>  I'm not sure a counter like you're suggesting is any clearer since the counter value would be reset anytime the server or agent was restarted, effectively making it a constant value of 1 for each active instance.

Hmm... it's my impression that the emitted value of a counter is the amount to increment by, rather than the absolute value. A little googling seems to support that, but maybe there is something I'm missing?

> The value-prop of a value of 1 is that you can use an aggregation function (such as sum) to get the count of instances, rather than "counting" the series.

Heh... the value I see in a counter is that you can easily compute the start/restart rate across a set of hosts. I don't think that can be done currently? Seems like it would be useful.

> Off the top of my head I'm not sure there exists a query language (such as promql, graphite, m3ql, etc.) that allows you to "count" series grouped by tags. For example, if I have the ability to query global metrics, and I want a count of running version by-deployment-area I think I need to use a sum / groupBy function, rather than a "count of series".

I see what you're saying... In the graphite docs, they have both a groupByTags function and countSeries function which looks like they can be combined to do something like this. That being said, the DataDog docs recommend a gauge for this use case. I guess I could go either way.

Reply:
:whynotboth: :trollface:

Reply:
Do we need to periodically re-emit the metric if we use a gauge? Some metrics systems I worked with in the past will only report the gauge value for a fixed number of minutes after it was emitted... for instance, a gauge that emits `1` once will appear with value 1 for five minutes from the time of emission, after which it will report null/no data.

Reply:
> Heh... the value I see in a counter is that you can easily compute the start/restart rate across a set of hosts. I don't think that can be done currently? Seems like it would be useful.

I've taken to emitting "milliseconds since start" to monitor for crash loops. 

```
func (r *uptimeReporter) start() {
	r.wg.Add(1)
	go func() {
		defer r.wg.Done()

		for {
			select {
			case <-r.ctx.Done():
				return
			case <-r.clk.After(_reportInterval):
				r.uptime.Update(float64(r.clk.Now().Sub(r.startTime) / time.Millisecond))
			}
		}
	}()
}
```
where `r.uptime` is a gauge. You can easily set a threshold along the lines of 'if any of my instances have been running for less than 1 minute over a 2 minute period; alarm'

Reply:
> Do we need to periodically re-emit the metric if we use a gauge? Some metrics systems I worked with in the past will only report the gauge value for a fixed number of minutes after it was emitted... for instance, a gauge that emits `1` once will appear with value 1 for five minutes from the time of emission, after which it will report null/no data.

That's not the experience I've had but we could periodically re-emit if needed by certain backends. Maybe this should be hidden behind the metrics facade?

Reply:
tally emits the current value at every reporting interval https://github.com/uber-go/tally/blob/master/stats.go#L127

Reply:
prom holds the current value in memory and will report it when scrapped https://github.com/armon/go-metrics/blob/master/prometheus/prometheus.go#L143

Reply:
None of the statsd docs seems to indicate that you must re-emit a gauge.

> Gauges are a constant data type. They are not subject to averaging, and they don’t change unless you change them. That is, once you set a gauge value, it will be a flat line on the graph until you change it again.

https://statsd.readthedocs.io/en/v3.2.1/types.html#gauges

Reply:
Thanks for all this supporting information @ZymoticB 

> I've taken to emitting "milliseconds since start" to monitor for crash loops.

That sounds much more effective than what I was suggesting.

With regards to emission frequency, I suppose we can always switch the behavior to emit more than once at some point in the future without affecting much (if we find that we need to).

Reply:
Makes sense, @ZymoticB . I think we are all in agreement that uptime is a really useful metric to emit. I can change this to be a gauge metric called "uptime" with a value of "milliseconds since startup" using the approach you suggested.

Reply:
I'm a bit confused by @rturner3 's comment so I want to clarify.

The suggestion to emit milliseconds since start to monitor for crash loops was in response to @evan2645 's request for a counter that could be used to monitor starts / stops. I agree we all see value in an uptime metric.

That said, IMO we should still have a gauge that emits version so that we can see "how many running instances are running version X", I don't think it should be a single uptime gauge.

Reply:
Thanks for clarifying, @ZymoticB . I think this thread diverged into multiple directions and the linear nature of GitHub conversations had me a little confused as to what exactly was being proposed for this PR.

I can create a separate issue for the uptime metric and keep this PR focused to just the version metric, which is currently implemented with a gauge.